Implemented cosine distance metric

dariocazzani opened the pull request:
Losses in `tf.contrib.losses.metric_learning` rely only on euclidean distance.
I implemented the cosine distance that can be used to calculate the losses for:

- `lifted_struct_loss`
- `cluster_loss`
- `triplet_semihard_loss`

I added an extra argument `metric` with default value `euclidean`, in order to make the API backward compatible.

dariocazzani committed:
Implemented cosine distance metric for triplet loss, cluster loss and contrastive loss

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -180,7 +208,12 @@ def triplet_semihard_loss(labels, embeddings, margin=1.0):
   # Build pairwise squared distance matrix.
-  pdist_matrix = pairwise_distance(embeddings, squared=True)
+  if metric == 'euclidean':
+    pdist_matrix = pairwise_distance(embeddings)
```
Don't change previous default behavior
squared=True

dariocazzani replied:
Right, sorry!

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -154,7 +182,7 @@ def masked_minimum(data, mask, dim=1):
 
 
-def triplet_semihard_loss(labels, embeddings, margin=1.0):
+def triplet_semihard_loss(labels, embeddings, margin=1.0, metric='euclidean'):
```
instead of an string pass a function, i.e. pairwise_dist_fn

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -404,7 +437,7 @@ def npairs_loss_multilabel(sparse_labels, embeddings_anchor,
 
 
-def lifted_struct_loss(labels, embeddings, margin=1.0):
+def lifted_struct_loss(labels, embeddings, margin=1.0, metric='euclidean'):
```
Same here

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -942,6 +980,7 @@ def func_body(iteration, gt_cluster_score):
 def cluster_loss(labels,
                  embeddings,
                  margin_multiplier,
+                 metric='euclidean'
```
Same here

dariocazzani committed:
Fixes after comments for PR:

- pairwise_distance_cosine function accepts 2 arguments. Second is a placeholder to make it compatible with pairwise_distance_cosine
- Isntead of passing a string to decide which metric to use, pass a function to the losses

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -80,7 +80,7 @@ def pairwise_distance(feature, squared=False):
   return pairwise_distances
 
-def pairwise_distance_cosine(feature):
+def pairwise_distance_cosine(feature, squared=True):
```
Add squared to the list of arguments, and say it is not used.

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -80,6 +80,34 @@ def pairwise_distance(feature, squared=False):
+  output[i, j] = 1 - tf.matmul(feature[i, :], feature[j, :]) / (tf.norm(feature[i, :]) * tf.norm(feature[j, :]))
+
+  Args:
+    feature: 2-D Tensor of size [number of data, feature dimension].
```
Add squared to the list arguments.

dariocazzani replied:
Done

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -154,7 +182,7 @@ def masked_minimum(data, mask, dim=1):
 
 
-def triplet_semihard_loss(labels, embeddings, margin=1.0):
+def triplet_semihard_loss(labels, embeddings, margin=1.0, pairwise_dist_fn=pairwise_distance):
```
Add pairwise_dist_fn to the list of args

dariocazzani replied:
Done

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -404,7 +432,7 @@ def npairs_loss_multilabel(sparse_labels, embeddings_anchor,
 
 
-def lifted_struct_loss(labels, embeddings, margin=1.0):
+def lifted_struct_loss(labels, embeddings, margin=1.0, pairwise_dist_fn=pairwise_distance):
```
Same here

dariocazzani replied:
Done

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:
```diff
@@ -942,6 +969,7 @@ def func_body(iteration, gt_cluster_score):
 def cluster_loss(labels,
                  embeddings,
                  margin_multiplier,
+                 pairwise_dist_fn=pairwise_distance,
```
Same here

dariocazzani replied:
Done

sguada commented on tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py:
```diff
@@ -57,6 +57,18 @@ def pairwise_distance_np(feature, squared=False):
+      dist_matrix = metric_loss_ops.pairwise_distances_cosine(embedding)
+      self.assertAllClose(dist_matrix.eval(), sklearn_result, rtol=1e-06)
+
 
```
Add a test where you change the pairwise_dist_fn

dariocazzani replied:
Done

dariocazzani committed:
Fixed arguments comments help

dariocazzani committed:
Implemented test for triplet_semihard_loss with pairwise distances calculated with cosine distance

dariocazzani committed:
Fixed typo

tensorflowbutler commented:
Nagging Reviewer @ispirmustafa, @sguada: It has been 15 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

dariocazzani commented:
Hi @case540 - I was wondering if not the PR can be merged.
Thanks

dariocazzani commented:
@kokoro-team 
I am not sure how to fix the checks that were not successful

case540 commented:
This seems like the relevant logs

==================== Test output for //bazel_pip/tensorflow/contrib/losses:metric_loss_ops_test:
  File "/home/kbuilder/.cache/bazel/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/execroot/org_tensorflow/bazel-out/k8-opt/bin/bazel_pip/tensorflow/contrib/losses/metric_loss_ops_test.runfiles/org_tensorflow/bazel_pip/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py", line 215
    pairwise_dist_fn=metric_loss_ops.pairwise_distances_cosine))
                                                               ^
SyntaxError: invalid syntax
================================================================================

dariocazzani committed:
Fixed typo

dariocazzani committed:
fixed improper use of tf.test

dariocazzani commented:
@case540 - Once again, I hope it's fixed now.

case540 commented:
Thanks for working through the errors. This seems now like most recent issues...


ERROR: testPairwiseCosine (__main__.PairwiseCosineTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/s/w/ir/run/bazel-out/k8-opt/bin/tensorflow/contrib/losses/metric_loss_ops_test.runfiles/org_tensorflow/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py", line 68, in testPairwiseCosine
    sklearn_result = 1 - metric.pairwise.cosine_similarity(embedding)
NameError: name 'metric' is not defined

ERROR: testTripletSemiHardCosine (__main__.TripletSemiHardLossCosineTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/s/w/ir/run/bazel-out/k8-opt/bin/tensorflow/contrib/losses/metric_loss_ops_test.runfiles/org_tensorflow/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py", line 176, in testTripletSemiHardCosine
    pdist_matrix = 1 - metric.pairwise.cosine_similarity(embedding)
NameError: name 'metric' is not defined

dariocazzani commented:
@case540 - Little by little I'll make this work and merged :)

dariocazzani commented:
@case540 
Now I am confused.
The error is now
```
AttributeError: module 'tensorflow.contrib.losses.python.metric_learning' has no attribute 'pairwise_distances_cosine'
```

But that function is the one I implemented and defined in `'tensorflow.contrib.losses.python.metric_learning'`.
I import it together with the rest: 
```python
from tensorflow.contrib.losses.python import metric_learning as metric_loss_ops
```

dariocazzani commented:
@case540 - ready to test again

dariocazzani commented:
@case540 
I need your help with this.
The error is
```bash
AttributeError: module 'tensorflow.contrib.losses.python.metric_learning' has no attribute 'pairwise_distance_cosine'
```
But `pairwise_distance_cosine` is defined in `tensorflow.contrib.losses.python.metric_learning` (it's the module where I added it)
I don't understand why `MacOS Contrib` doesn't fail and `Ubuntu contrib` does.

Thanks

case540 commented:
Guessing you have to modify tensorflow/contrib/losses/python/metric_learning/__init__.py to add the API to _allowed_symbols? Not really familiar with the code too much

case540 commented:
Also, you will need to fix these pylint errorrs....


FAIL: Found 4 non-whitelited pylint errors:
tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:86: [C0301(line-too-long), ] Line too long (112/80)
tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:97: [C0326(bad-whitespace), ] Exactly one space required around assignment
tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py:186: [C0301(line-too-long), ] Line too long (94/80)
tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py:179: [C0330(bad-continuation), ] Wrong hanging indentation (add 4 spaces).

And sorry for taking so long to respond

dariocazzani committed:
Fix allowed_symbols and pylint errors

dariocazzani commented:
@case540 
I fixed the `_allowed_symbols`, but I am not sure that was the problem.
Can you run the checks? If what I did doesn't solve it, maybe someone else can pitch in?

Thanks

tensorflowbutler commented:
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

case540 commented:
Sorry for the delay. There are still test failures. Rerunning tests, but please take a look at failures if they still exist.

tensorflowbutler commented:
It has been 20 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

tensorflowbutler commented:
It has been 49 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
